src/node_api.cc

```diff
@@ -574,16 +681,14 @@ napi_status napi_open_callback_scope(napi_env env,
 
   v8::Local<v8::Context> context = env->context();
 
-  node::async_context* node_async_context =
-      reinterpret_cast<node::async_context*>(async_context_handle);
+  v8impl::AsyncContext* node_async_context =
+      reinterpret_cast<v8impl::AsyncContext*>(async_context_handle);
 
   v8::Local<v8::Object> resource;
   CHECK_TO_OBJECT(env, context, resource, resource_object);
 
   *result = v8impl::JsCallbackScopeFromV8CallbackScope(
-      new node::CallbackScope(env->isolate,
-                              resource,
-                              *node_async_context));
+      new v8impl::AsyncContextCallbackScope(node_async_context));
```

Comment:
The `resource_object` passed to `napi_open_callback_scope` is no long used here instead that one from `async_context_handle` is used. Not sure if such a behavior change is acceptable.

Reply:
I don't remember any history which would explain why we have the napi_value async_resource parameter in napi_open_callback_scope when the async_resource parameter is also in async_init.  

We should look at adding a new method without that parameter which operates as suggested here and deprecate the one with the method but not remove it.  It would however only be available in a newer N-API version.

The remaining question on the Async Hooks side is if there is any sensible reason you would want to override the resource_object in the async_context passed in since the existing method allows people to do that? If not, then since Async Hooks is still an experimental feature are we comfortable with changing the behavior for all users? If yes then we need to update the documentation for the say the parameter is ignored.  If the answer is no, then we could probably update the docs to say that you should only pass in NULL and that in this case the resource object from the context will be used and update the code to do that.

@Flarna thoughts?

Reply:
I don't believe changing the resource object of a definite async id is acceptable according to current async hooks API like `executionAsyncResourc`, upon which we have built `AsyncLocalStorage` to propagate local storages across async execution contexts.

Reply:
@mhdawson I was also not that involved in NAPI API so I can mostly guess here.
In the initial NAPI design async-hooks were not included at all, e.g. `napi_make_callback` in 8.0.0 had no `async_context` parameter and the `napi_async_XXX` APIs did not exist.
At the time NAPI was made async-hooks aware the resource was most likely not seen as "important"/"relevant". It's still allowed to pass `NULL` and it's not ensured that async_id matches.
Most likely the idea was to stay flexible. Even async-hooks doc of today states `In some cases the resource object is reused for performance reasons, it is thus not safe to use it as a key in a WeakMap or add properties to it.` which is a clearly a problem for `executionAsyncResource`.
async-hooks are still experimental and in flow...

In short based on the latest changes in async-hooks (add `executionAsyncResource`,...) it's needed to ensure that async_id matches resource and as a result both are unique.
With this in mind the changes here look fine - they could be even seen as bugfix. But I'm not sure about the doc process in NAPI. NAPI is not experimental so changing the semantics of an API may cause issues for some people (not for me).
Besides that from API surface point of view we should also think about getting rid of ignored parameters.

Reply:
Based on the discussion my proposal would be

1) We adjust the functionality as proposed in the PR since async hooks is experimental for the existing function. In addition we update the doc to indicated that the resource_object is ignored.
2) we add a new function which drops the resource_object parameter, and doc deprecate the existing method. (we can't necessarily remove it due to the N-API ABI guarantees).

Reply:
👍 nice wrapping up! IIUC, either or not we are going to deprecate the existing function and propose a new one, we'll have to determine if it is acceptable to make the resource_object parameter a noop parameter and update the docs accordingly.

Reply:
Maybe we should also think about adapting `napi_async_init` as it states that the resource is optional. As Node is meanwhile built to require an async resource it seems to be strange that we create an object on the fly if user doesn't provide one.
Removing the null check in `napi_async_init` would be no issue from ABI point of view but breaking from behavior. Not sure if this should be included in this PR.

Reply:
@Flarna I'd suggest we break that out into a different issue.

Reply:
@Flarna @mhdawson I created #33153 to track the semantic of `napi_async_init`.